src/movies.js

```diff
@@ -1,16 +1,249 @@
 // Iteration 1: All directors? - Get the array of all directors.
+
+function getAllDirectors(arrayMovies) {
+    const directorArray = arrayMovies.map(function(movie) {
+        return movie.director;
+    })
+        return directorArray;
+    }
 // _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the array of directors. How could you "clean" a bit this array and make it unified (without duplicates)?
 
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
+function howManyMovies(array1) {
+
+    if (array1.length === 0) {
+        return 0;
+      }
 
+    const spielbergDrama = array1.filter(function(movie) {
+      //checks to see if directore is Steven
+      if(movie.director ==="Steven Spielberg") {
+          //iterates over the genre key to see if drama
+          for(let i = 0; i < movie.genre.length; i ++) {
+            //returns movie if it is a drama
+            if(movie.genre[i] === "Drama" ) {
+              return movie;
+            }
+          }
+      }
+  
+    })
+    return spielbergDrama.length;
+  }
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
+function ratesAverage(array1) {
 
+    if (array1.length === 0) {
+      return 0;
+    }
+  
+  const movieRating = array1.reduce(function(movieRating, current){
+    
+       return movieRating + current.rate;
+ 
+  
+  },0);
+  let avgRating = movieRating / array1.length;
+  
+  return Number(avgRating.toFixed(2));
+  }
+  
 // Iteration 4: Drama movies - Get the average of Drama Movies
+function dramaMoviesRate(array) {
+
+    if (array.length === 0) {
+      return 0;
+    }
+    //this array filters for movies that are drama and returns it
+    const onlyDramaMovies = array.filter(function(movie) {
+      if(movie.genre.indexOf('Drama') >= 0) {
+        return movie;
+      }
+    });
+    //tests to see if there are no drama movies
+
+    if (onlyDramaMovies.length === 0) return 0;
+  
+    //this array reduces the drama movie ratings into onlyDramaMovies
+    const dramaRating = onlyDramaMovies.reduce(function(accumu, currValue) {
+        return accumu + currValue.rate;
+  
+    }, 0);
+    let numOfDramaMovies = onlyDramaMovies.length;
+    
+    let avgRating = dramaRating / numOfDramaMovies;
+  
+   return Number(avgRating.toFixed(2));
+  }
+  
 
 // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
+function orderByYear(array) {
+    if (array.length === 0) {
+        return 0;
+      }
+    
+      const sortedByYear = array.sort(function (movie1, movie2) {
```

Comment:
`.sort()` method will mutate your original array so, to avoid that, we should first make a copy of the original array:

```js
function orderByYear(array) {
  const newArr = [...array];
  newArr.sort(...)
}
```